From 2bc1def96413d93e7689559b3bd5b95307c270a5 Mon Sep 17 00:00:00 2001
From: Greg Hudson <ghudson@mit.edu>
Date: Mon, 12 Aug 2013 15:17:20 -0400
Subject: [PATCH] Fix localauth memory leak

localauth modules were not freed by krb5_free_context(), causing a
memory leak.
---
 src/lib/krb5/krb/init_ctx.c  | 2 ++
 src/lib/krb5/libkrb5.exports | 1 +
 src/lib/krb5/os/localauth.c  | 7 +++++++
 src/lib/krb5/os/os-proto.h   | 1 +
 4 files changed, 11 insertions(+)

diff --git a/src/lib/krb5/krb/init_ctx.c b/src/lib/krb5/krb/init_ctx.c
index 059f764..499f9f0 100644
--- a/src/lib/krb5/krb/init_ctx.c
+++ b/src/lib/krb5/krb/init_ctx.c
@@ -51,6 +51,7 @@
 
 #include "k5-int.h"
 #include "int-proto.h"
+#include "os-proto.h"
 #include <ctype.h>
 #include "brand.c"
 #include "../krb5_libinit.h"
@@ -289,6 +290,7 @@ krb5_free_context(krb5_context ctx)
 #endif
 
     k5_ccselect_free_context(ctx);
+    k5_localauth_free_context(ctx);
     k5_plugin_free_context(ctx);
     free(ctx->plugin_base_dir);
 
diff --git a/src/lib/krb5/libkrb5.exports b/src/lib/krb5/libkrb5.exports
index 00dc087..258f481 100644
--- a/src/lib/krb5/libkrb5.exports
+++ b/src/lib/krb5/libkrb5.exports
@@ -113,6 +113,7 @@ initialize_prof_error_table
 k5_ccselect_free_context
 k5_free_serverlist
 k5_kt_get_principal
+k5_localauth_free_context
 k5_locate_kdc
 k5_plugin_free_modules
 k5_plugin_load
diff --git a/src/lib/krb5/os/localauth.c b/src/lib/krb5/os/localauth.c
index e48b3a9..921b654 100644
--- a/src/lib/krb5/os/localauth.c
+++ b/src/lib/krb5/os/localauth.c
@@ -453,3 +453,10 @@ krb5_aname_to_localname(krb5_context context, krb5_const_principal aname,
     }
     return KRB5_LNAME_NOTRANS;
 }
+
+void
+k5_localauth_free_context(krb5_context context)
+{
+    free_handles(context, context->localauth_handles);
+    context->localauth_handles = NULL;
+}
diff --git a/src/lib/krb5/os/os-proto.h b/src/lib/krb5/os/os-proto.h
index 1ef94a2..04603c5 100644
--- a/src/lib/krb5/os/os-proto.h
+++ b/src/lib/krb5/os/os-proto.h
@@ -115,6 +115,7 @@ extern unsigned int krb5_skdc_timeout_shift;
 extern unsigned int krb5_skdc_timeout_1;
 extern unsigned int krb5_max_dgram_size;
 
+void k5_localauth_free_context(krb5_context);
 krb5_error_code localauth_names_initvt(krb5_context context, int maj_ver,
                                        int min_ver, krb5_plugin_vtable vtable);
 krb5_error_code localauth_rule_initvt(krb5_context context, int maj_ver,
-- 
2.6.2

